Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
-}
+class __UserDataProvider : public __CGDataProviderInternal {
+    mutable void* m_info;
+    const CGDataProviderCallbacks* m_callbacks;
+    mutable std::vector<unsigned char> m_data;
 
-- (NSUInteger)length {
-    return [_data length];
-}
+public:
+    __UserDataProvider(void* info, const CGDataProviderCallbacks* callbacks) : m_info(info), m_callbacks(callbacks) {
+    }
 
-@end
+    const void* GetData() const override {
+        if (m_data.empty()) {
+            size_t count = 0, delta = 0;
+            do {
+                m_data.resize(m_data.size() + c_streamSize);
+                delta = m_callbacks->getBytes(m_info, m_data.data() + count, c_streamSize);
+                count += delta;
+                m_callbacks->skipBytes(m_info, delta);
+            } while (delta == c_streamSize);
+            m_data.resize(count);
+        }
 
-/**
- @Status Caveat
- @Notes Only supports file:/// URLs
-*/
-CGDataProviderRef CGDataProviderCreateWithURL(CFURLRef url) {
-    NSString* path = [static_cast<NSURL*>(url) path];
-    CGDataProvider* ret = [[CGDataProvider alloc] initWithContentsOfFile:path];
-    if (ret) {
-        ret->filename = path;
+        return m_data.data();
     }
 
-    return ret;
-}
+    size_t GetSize() const override {
+        if (m_data.empty()) {
+            GetData();
+        }
+
+        return m_data.size();
+    }
+
+    ~__UserDataProvider() {
+        m_callbacks->releaseProvider(m_info);
+    }
+};
+
+class __RawDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    const void* m_data;
+    size_t m_size;
+    CGDataProviderReleaseDataCallback m_releaseData;
+
+public:
+    __RawDataProvider(void* info, const void* data, size_t size, CGDataProviderReleaseDataCallback releaseData)
+        : m_info(info), m_data(data), m_size(size), m_releaseData(releaseData) {
+    }
+
+    const void* GetData() const override {
+        return m_data;
+    }
+
+    size_t GetSize() const override {
+        return m_size;
+    }
+
+    ~__RawDataProvider() {
+        m_releaseData(m_info, m_data, m_size);
+    }
+};
+
+class __DirectDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    off_t m_size;
+    const CGDataProviderDirectCallbacks* m_callbacks;
+    mutable const void* m_data = nullptr;
+
+public:
+    __DirectDataProvider(void* info, off_t size, const CGDataProviderDirectCallbacks* callbacks)
+        : m_info(info), m_size(size), m_callbacks(callbacks) {
+    }
+    const void* GetData() const override {
+        if (m_data == nullptr) {
+            m_data = m_callbacks->getBytePointer(m_info);
+        }
+
+        return m_data;
+    }
+
+    size_t GetSize() const override {
+        return m_size;
+    }
+
+    ~__DirectDataProvider() {
+        if (m_data != nullptr) {
+            m_callbacks->releaseBytePointer(m_info, m_data);
+        }
+
+        m_callbacks->releaseInfo(m_info);
+    }
+};
+
+class __DirectAccessDataProvider : public __CGDataProviderInternal {
```

Comment:
Unfortunately the types and names of the callbacks in the struct are different, making templatizing this more trouble than it's worth.  Ends up making the code a lot more complex imo.